templates/accounts/volunteer_profile.html

```diff
@@ -167,6 +167,32 @@ <h2 class="font-weight-bolder">Complete Your Profile</h2>
 
 
 {% block scripts %}
+    <!--  script for AJAX call -->
+    <script>
+        $("#rollNo").focusout(function () {
+            $.ajax({
```

Comment:
We also need to call this AJAX on clicking the `Continue` button as it may happen that the user fill the roll number field at the last and submit the form (and hence `focusOut` event is not fired and if the roll number turns out to be duplicate, the user will lose all the entered data.

Possible solution is to take this AJAX call out as a function use it at both the places.

Reply:
On thinking about this for a while, I don't think it would be feasible to do AJAX call after submit button is clicked as AJAX calls are asynchronous and we can't stop the form from being submitted until we get a result from AJAX.

This only solution to this, as I think, would be to send AJAX request on `keyUp` event rather than `focusOut`. As all these requests are asynchronous, it won't affect the performance of the website and as the roll numbers are not too long to that won't make up too many requests to the server.

But, one thing to take care of is that we might need to add a small `sleep` time before sending AJAX requests as if any AJAX response gets delayed for some reason and the response to the request sent afterwards is received first, the former response should not override the later response.

Reply:
> On thinking about this for a while, I don't think it would be feasible to do AJAX call after submit button is clicked as AJAX calls are asynchronous and we can't stop the form from being submitted until we get a result from AJAX.
> 
> This only solution to this, as I think, would be to send AJAX request on `keyUp` event rather than `focusOut`. As all these requests are asynchronous, it won't affect the performance of the website and as the roll numbers are not too long to that won't make up too many requests to the server.
> 
> But, one thing to take care of is that we might need to add a small `sleep` time before sending AJAX requests as if any AJAX response gets delayed for some reason and the response to the request sent afterwards is received first, the former response should not override the later response.

but it will increase the load on the database unnecessarily. It might be overkilling for this problem.

Reply:
> We also need to call this AJAX on clicking the `Continue` button as it may happen that the user fill the roll number field at the last and submit the form (and hence `focusOut` event is not fired and if the roll number turns out to be duplicate, the user will lose all the entered data.
> 
> Possible solution is to take this AJAX call out as a function use it at both the places.

Not exactly. as soon as you click on submit button, focusout event happens and the AJAX call will be made.

Reply:
But before the AJAX call is completed, the form is already submitted.

To reproduce, first fill the form with correct roll number and then at last, fill the wrong roll number and click `Continue`.

Reply:
> But before the AJAX call is completed, the form is already submitted.
> 
> To reproduce, first fill the form with correct roll number and then at last, fill the wrong roll number and click `Continue`.

Got that problem.

Reply:
> On thinking about this for a while, I don't think it would be feasible to do AJAX call after submit button is clicked as AJAX calls are asynchronous and we can't stop the form from being submitted until we get a result from AJAX.
> 
> This only solution to this, as I think, would be to send AJAX request on `keyUp` event rather than `focusOut`. As all these requests are asynchronous, it won't affect the performance of the website and as the roll numbers are not too long to that won't make up too many requests to the server.
> 
> But, one thing to take care of is that we might need to add a small `sleep` time before sending AJAX requests as if any AJAX response gets delayed for some reason and the response to the request sent afterwards is received first, the former response should not override the later response.

Rather than putting some fixed time, I would suggest to use synchronous calls for AJAX requests.

Reply:
Yeah, that'll be better. You can create a separate function for AJAX call with a parameter `async_call`. For `focusOut` event, the argument may be passed as `true` while `false` in case of clicking on submit button.

The AJAX function in form submission case can be called after this line.
https://github.com/garg3133/JagratiWebApp/blob/c993b2ba7fe5423da780682f125c286511baae11/templates/accounts/volunteer_profile.html#L179

Reply:
As you said, we are not supposed to stop form submission for an AJAX request.